examples/reshare/src/application/actor.rs

```diff
@@ -0,0 +1,235 @@
+use crate::{
+    application::{types::genesis_block, Block, Mailbox, Message},
+    dkg,
+    utils::{get_last_height, height_in_epoch, BLOCKS_PER_EPOCH},
+};
+use commonware_consensus::{marshal, types::Round};
+use commonware_cryptography::{
+    bls12381::primitives::variant::Variant, Committable, Digestible, Hasher, Signer,
+};
+use commonware_runtime::{spawn_cell, Clock, ContextCell, Handle, Metrics, Spawner};
+use futures::{
+    channel::mpsc,
+    future::{try_join, Either},
+    lock::Mutex,
+    StreamExt,
+};
+use rand::Rng;
+use std::{future, sync::Arc, time::Duration};
+use tracing::{info, warn};
+
+/// The application [Actor].
+pub struct Actor<E, H, C, V>
+where
+    H: Hasher,
+    C: Signer,
+    V: Variant,
+{
+    context: ContextCell<E>,
+    mailbox: mpsc::Receiver<Message<H, C, V>>,
+}
+
+impl<E, H, C, V> Actor<E, H, C, V>
+where
+    E: Rng + Spawner + Metrics + Clock,
+    H: Hasher,
+    C: Signer,
+    V: Variant,
+{
+    /// Create a new application [Actor] and its associated [Mailbox].
+    pub fn new(context: E, mailbox_size: usize) -> (Self, Mailbox<H, C, V>) {
+        let (sender, mailbox) = mpsc::channel(mailbox_size);
+
+        (
+            Self {
+                context: ContextCell::new(context),
+                mailbox,
+            },
+            Mailbox::new(sender),
+        )
+    }
+
+    /// Start the application.
+    pub fn start(
+        mut self,
+        marshal: marshal::Mailbox<V, Block<H, C, V>>,
+        dkg: dkg::Mailbox<H, C, V>,
+    ) -> Handle<()> {
+        spawn_cell!(self.context, self.run(marshal, dkg).await)
+    }
+
+    /// Application control loop
+    async fn run(
+        mut self,
+        mut marshal: marshal::Mailbox<V, Block<H, C, V>>,
+        dkg: dkg::Mailbox<H, C, V>,
+    ) {
+        let genesis = genesis_block();
+        let genesis_digest = genesis.digest();
+        let built = Arc::new(Mutex::new(None));
+
+        while let Some(message) = self.mailbox.next().await {
+            match message {
+                Message::Genesis { epoch, response } => {
+                    // Case: Genesis.
+                    if epoch == 0 {
+                        let _ = response.send(genesis_block::<H, C, V>().commitment());
+                        continue;
+                    }
+
+                    // Case: Non-genesis.
+                    let height = get_last_height(epoch - 1);
+                    let Some(block) = marshal.get_block(height).await else {
+                        // No block exists, put the response in the responders map for later.
```

Comment:
Still reviewing but in `main` I think consensus will panic if you continue here rather than waiting?

Reply:
Correct - made this panic instead. We'll never start a new epoch's consensus engine without having the genesis.